Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,121 @@
+# DaemonSet deployment for Prometheus Agent
+
+* Owners:
+  * [haanhvu](https://github.com/haanhvu)
+* Related Tickets:
+  * [#5495](https://github.com/prometheus-operator/prometheus-operator/issues/5495)
+* Other docs:
+  * n/a
+
+This proposal is about designing and implementing the deployment of Prometheus Agent as DaemonSet. Currently, Prometheus Agent can only be deployed as StatefulSet, which could be considered as “cluster-wide” strategy, meaning one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster. Though this works well for many use cases, some use cases may indeed prefer a “node-specific” strategy (DaemonSet), where each Prometheus Agent only scrapes the metrics from the targets located on the same node.
+
+
+## 1. Why
+
+When deploying Prometheus Agent in Kubernetes, two of the biggest users’ concerns are load distribution and scalability.
+
+DaemonSet deployment is a good solution for these:
+* Load distribution: Each Prometheus Agent pod will only scrape the targets located on the same node. Even though the targets on some nodes may produce more metrics than other nodes, the load distribution is reliable enough. This has been proven in [Google Cloud Managed Service for Prometheus (GMP)'s operator](https://github.com/GoogleCloudPlatform/prometheus-engine/) which follows a similar approach.
+* Automatic scalability: When new nodes are added to the cluster, new Prometheus Agent pods will be automatically added in those nodes. Users can also select which set of nodes they want to deploy Prometheus Agent and the priority of Prometheus Agent pods compared to other pods on the same node.
+
+DaemonSet deployment is especially more suitable for Prometheus Agent than Prometheus, since the Agent mode is more lightweight. In specific, it requires (currently around 20 to 30%) less memory, does not produce TSDB blocks on disks, and naturally blocks querying APIs.
```

Comment:
It is kinda implicit but I'd emphasize that Prometheus Agent with daemonset is also best suited for a collect-and-forward approach: when you push all metrics to a central datastore which can scale horizontally itself, you don't have to care about query and rule distribution which is the main issue when sharding Prometheus servers.

Reply:
> you don't have to care about query and rule distribution which is the main issue when sharding Prometheus servers.

Good callout. With recording/alerting rules are disabled in agent mode. how should the operator handle reconciling existing `PrometheusRule` resources?

It looks like Prometheus will [error](https://github.com/prometheus/prometheus/blob/1081e336a0c4004b6f817c70f0ff4f3810db80fa/config/config.go#L124) on config reloading if they are present.

Apologies if this is too much into the details. I haven't reviewed a PO proposal before and I am unfamiliar with how much is covered at this level 😄

Reply:
Currently, changes to PrometheusRule don't trigger reconciliations of PrometheusAgent CRDs 🙂  (We have separate CRDs for Prometheus and PrometheuAgents)

Reply:
Ahhh - I didn't realize that. Thanks for clarifying!